File: 2026-02-February-GFG-POTD/February 01 - K Sized Subarray Maximum/Solution.cpp
class Solution {
public:
    vector<int> maxOfSubarrays(vector<int>& arr, int k) {
        vector<int> result;
        deque<int> dq;  // stores indices

        for (int i = 0; i < arr.size(); i++) {

            // Remove indices out of current window
            while (!dq.empty() && dq.front() <= i - k) {
                dq.pop_front();
            }

            // Remove smaller elements from the back
            while (!dq.empty() && arr[dq.back()] <= arr[i]) {
                dq.pop_back();
            }

            // Add current index
            dq.push_back(i);

            // Store max for the window
            if (i >= k - 1) {
                result.push_back(arr[dq.front()]);
            }
        }
        return result;
    }
};

File: 2026-02-February-GFG-POTD/February 01 - K Sized Subarray Maximum/Solution.py
from collections import deque

class Solution:
    def maxOfSubarrays(self, arr, k):
        result = []
        dq = deque()  # stores indices

        for i in range(len(arr)):

            # Remove indices out of current window
            while dq and dq[0] <= i - k:
                dq.popleft()

            # Remove smaller elements from the back
            while dq and arr[dq[-1]] <= arr[i]:
                dq.pop()

            # Add current index
            dq.append(i)

            # Store max for the window
            if i >= k - 1:
                result.append(arr[dq[0]])

        return result
